### ISSUE TITLE
Forks,forks,forks

### ISSUE THREAD

**Issue by gropox** [GolosChain/golos#271]
Last time very often in logs appearing follong messages

```
3150271ms th_a       database.cpp:3641             update_last_irrevers ] Encountered a block num collision due to a fork. Walking the current fork to determine the correct block. block_num:11206696
3321057ms th_a       database.cpp:3641             update_last_irrevers ] Encountered a block num collision due to a fork. Walking the current fork to determine the correct block. block_num:11206748
3516026ms th_a       database.cpp:3641             update_last_irrevers ] Encountered a block num collision due to a fork. Walking the current fork to determine the correct block. block_num:11206814
3525018ms th_a       database.cpp:3641             update_last_irrevers ] Encountered a block num collision due to a fork. Walking the current fork to determine the correct block. block_num:11206820

```

Affected are all nodes I running. Namely it is coming last time very often to problems with scripts. E.g. transfers go into nowhere or scripts break with ```Cannot read property 'previous' of null``` because golos-js library cannot read last block. It is almost impossible to run reliably the scripts.

It seems also to be the cause for very often missing blocks by witnesses. 

Does some one running nodes  with HF17 version live, on network? Devs? 

How could this issue be investigated?

**Comment by kotbegemot** [comment]
Multiple usage of a witness plug-in with a private key leads to the problems described above.

**Comment by kotbegemot** [comment]
Where did you get this problem? (Production)

**Comment by gropox** [comment]
Yes, 0.16.4 It is on all my nodes. Some witnesses reported this problem also. 

![image](https://user-images.githubusercontent.com/26039472/32704513-9accc4e4-c806-11e7-9cba-21bb293d9ee0.png)

**Comment by gropox** [comment]
I nothing changed in configuration of my witness node, last 3-4 months. It is appearing last time, since september-october

**Comment by gropox** [comment]
![image](https://user-images.githubusercontent.com/26039472/32704524-d8ecb19e-c806-11e7-831f-13bc89ff3abd.png)

**Comment by kotbegemot** [comment]
The number of transactions has increased in the network.
You started to catch the problem with a fork.
You can not usage a multiple delegate plug-in with the same private key.

**Comment by gropox** [comment]
> The number of transactions has increased in the network.
> You started to catch the problem with a fork.

Not only me get these errors, but also other witnesses! Or do you mean, we reached performance limits of golos network? Then how we can work on such blockchain, if it is not predictably which transaction leads to an error? Do you want to do something against that? Or how should i understand your sentence? 

> You can not usage a multiple witness plug-in with the same private key.

Once again, actually I have only one witness node, and three full nodes for my scripts. All nodes experiencing  this problem, and not only by my nodes!

In summer there wasn't such problems at all.

**Comment by Nemo1369** [comment]
@gropox @kotbegemot Well, I think that is enough. Needs to be explained well. 

These trouble seems to be occurring when someone is running duplicate witness node or (what is less probable) a fork daemon. As far as you understand, we have not changed the production daemon codebase yet, but you can easily reproduce the issue with running test network with duplicate witness nodes. Well, I'm going to find out who is this duplicate-node person.

**Comment by gropox** [comment]
The problem seems to be gone. I have these Last block from 2017-11-13T09:28:57 GMT timezone. Probably someone turned off backup node around this time. 

```
1977058ms th_a       database.cpp:3641             update_last_irrevers ] Encountered a block num collision due to a fork. Walking the current fork to determine the correct block. block_num:11219426                      
1785074ms th_a       database.cpp:3641             update_last_irrevers ] Encountered a block num collision due to a fork. Walking the current fork to determine the correct block. block_num:11220561
``` 

Could be possible to build in a better logging for such cases? witness name at least? Then we could identify the witness sooner.

**Comment by vikxx** [comment]
During --replay, one of nodes is stuck at the moment with warn, which has been repeated for ~20 min
"Pushing block to fork database that failed to link"
```
1321958ms th_a       fork_database.cpp:39          push_block           ] Pushing block to fork database that failed to link: 00903ec83dbdc4e1c28914ad90ccd3c9154dd272, 9453256
1321958ms th_a       fork_database.cpp:40          push_block           ] Head: 9453249, 00903ec110fe0b7a11af346d5b45da1a2c4e19dc

```
Head num always: 9453249

**Comment by inertia186** [comment]
Is this related?

```bash
docker run \
    -d -p 2001:2001 -p 8090:8090 --name golos-full \
    goloschain/golos

docker logs -f golos-full
```

```
1950811ms p2p        node.cpp:2126                 on_hello_message     ] Received hello message from peer running a node for different blockchain.
```

**Comment by vikxx** [comment]
В коллекцию
![image](https://user-images.githubusercontent.com/26347159/33214069-738e5d9a-d0ce-11e7-9feb-78c98df4f872.png)

**Comment by vikxx** [comment]
https://youtu.be/4PiOqJPjUMs?t=179

**Comment by gropox** [comment]
Seems to be really duplicated signed by same witness blocks. In all 3 cases last 2 hours caused by miners. It has to be solved in a way soon. Because it is impossible to make reliably business using golos blockchain. After HF there a plannings to make business on golos, but it could be difficult. Blocks getting lost, transfers, posts, votes and so on. Newly was the case. Script made a transfer to the booster, but transferred amount didn't reach the target. Following the booster didn't vote for desired post. The money is not lost, but there is a a possible win loss. This is only one example. 

![image](https://user-images.githubusercontent.com/26039472/33998595-4f5879a4-e0e8-11e7-9c16-5a5bdb22ff17.png)

**Comment by gropox** [comment]
![image](https://user-images.githubusercontent.com/26039472/33999553-104d6064-e0eb-11e7-9652-408d6c101d15.png)

**Comment by gropox** [comment]
I would propose to disable accounts either by changing empty signing key or make it impossible for account to sign blocks for some time. For one week for example.

**Comment by gropox** [comment]
Two different witnesses!

![image](https://user-images.githubusercontent.com/26039472/34008752-aeb6f196-e106-11e7-89b4-0b350956c1bd.png)

**Comment by gropox** [comment]
I collected a list of accounts caused collisions today. About 101 cases. 101 lost blocks with data. The most of them are miners and they sign the block duplicated. The witnesses in the list seems to have other reasons for collisions. Probably timeouts or network issues. 

We do not have instruments to combat this. We could try to disable such miners/witnesses who are running duplicate nodes with same key. But by such amount of miners it will not help probably.

Thus I think we should disable the mining as soon as possible. Probably with a minor HF 16.5 in between.

```
      1 baybak
      1 bla1
      1 bla13
      1 bla14
      1 bla18
      1 bla19
      1 bla20
      1 bla21
      1 bla22
      1 bla23
      1 bla24
      1 bla26
      1 bla33
      1 bla40
      1 bla5
      1 bla8
      1 grub10
      1 grub13
      1 grub14
      1 grub16
      1 grub20
      1 grub21
      1 grub23
      1 grub25
      1 grub26
      1 grub29
      1 grub3
      1 grub8
      1 grub9
      1 grud10
      1 grud1
      1 grud4
      1 grud5
      1 grud6
      1 mway1
      1 mway28
      1 mway29
      1 mway3
      2 bla
      2 bla30
      2 bla32
      2 bla38
      2 grub15
      2 grub17
      2 grub22
      2 grub27
      2 grub28
      2 grub30
      2 grub4
      2 grud
      2 grud3
      2 grud7
      2 jesta
      2 mway23
      3 bla34
      3 grub
      3 grub1
      3 grub12
      3 xanoxt
      4 bla17
      4 grub11
      4 grub5
      4 lehard
```

**Comment by gropox** [comment]
For collection the infos, I added log outputs like that

![image](https://user-images.githubusercontent.com/26039472/34058916-ff9f2520-e1dc-11e7-98ab-2b5a98d012b1.png)

There is sample output of miners and witnesses

![image](https://user-images.githubusercontent.com/26039472/34058950-24e0fe08-e1dd-11e7-8388-17e722abc25f.png)

**Comment by vikxx** [comment]
> These trouble seems to be occurring when someone is running duplicate witness node

Problem is that you can specify the login of another witness in the config and harm him. 
That's, there is no duplicate private key, but **JUST DUPLICATE LOGIN**.

I put experiments on @VIM-Arcange as the best witness (who does't miss blocks long time before)
![image](https://user-images.githubusercontent.com/26347159/34146337-b1107b8a-e43d-11e7-8765-d69dad770ff4.png)

And his witness start missing blocks ...

![image](https://user-images.githubusercontent.com/26347159/34146391-ea940138-e43d-11e7-990f-efd365433b1e.png)

To exclude coincidences, now I will switch to another witness login (@gropox )
If ropox start missing blocks, and Arcange stop misses - it means that golosd allows harming another witness by just specifying his login.

Thus, our own backup nodes aren't important with which key, but it's important with what login.

For example, I stopped skipping blocks when I removed my logins from 3 backup nodes. 

![image](https://user-images.githubusercontent.com/26347159/34146671-023e67aa-e43f-11e7-884a-9f37f943821b.png)

**Comment by VIM-Arcange** [comment]
@vikxx Do each of your backup nodes have different private keys?

**Comment by vikxx** [comment]
>Do each of your backup nodes have different private keys?
#### Of course ! 

Moreover, I don't know your key or @gropox  key at all. And as I wrote above:

>To exclude coincidences, now I will switch to another witness login (@gropox )
If ropox start missing blocks, and Arcange stop misses - it means that golosd allows harming another witness by just specifying his login.

...and here we go:

![image](https://user-images.githubusercontent.com/26347159/34172430-20bb7464-e496-11e7-930c-7307c0495cd1.png)

**Comment by vikxx** [comment]
> Do each of your backup nodes have different private keys?

Of course
